plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -312,8 +282,8 @@ object StmtConversionVisitor : FirVisitor<ExpEmbedding, StmtConversionContext>()
         return when (typeOperatorCall.operation) {
             FirOperation.IS -> Is(argument, conversionType)
             FirOperation.NOT_IS -> Not(Is(argument, conversionType))
-            FirOperation.AS -> Cast(argument, conversionType).withAccessInvariants()
-            FirOperation.SAFE_AS -> SafeCast(argument, conversionType).withAccessInvariants()
+            FirOperation.AS -> Cast(argument, conversionType).withAccessAndProvenInvariants()
+            FirOperation.SAFE_AS -> SafeCast(argument, conversionType).withAccessAndProvenInvariants()
```

Comment:
Before there were only `withAccessInvariants()` calls. Because of that Viper could not prove that values returned by some of the methods are of the necessary type (we didn't see these issues before because those values were of special types like `Int`).

Reply:
I see.  I think a design decision is needed here: are we going to put these things in the predicates (in which case they'll be subsumed by @francescoo22 's changes), or will we keep inhaling them separately?  Could you discuss with Francesco about what is most convenient here?